File: java-impl/src/test/java/net/terria/registry/terraform/RegistryServiceTest.java
package net.terria.registry.terraform;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import static org.assertj.core.api.Assertions.assertThat;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;

@SpringBootTest
public class RegistryServiceTest {

    private static final Path REGISTRY_STORAGE_PATH = Path.of("../registry/");

    @Autowired
    private RegistryService registryService;

    @DynamicPropertySource
    static void dynamicProperties(DynamicPropertyRegistry registry) {
        registry.add("registry.storage.path", () -> REGISTRY_STORAGE_PATH.toString());
    }
    
    public static void setup() {

        System.out.println("Registry Storage Root: " + REGISTRY_STORAGE_PATH.toString());

        // Ensure registry storage directory exists before tests
        try {
            Files.createDirectories(REGISTRY_STORAGE_PATH);
        } catch (IOException e) {
            throw new RuntimeException("Failed to create registry storage directory", e);
        }
    }


    @BeforeEach
    public void initStorage() throws IOException {

        Path registryPath = REGISTRY_STORAGE_PATH.resolve("terraform");
        Path providersPath = registryPath.resolve("providers");
        Path modulesPath = registryPath.resolve("modules");

        Files.createDirectories(providersPath);
        Files.createDirectories(modulesPath);

        // Re-init service after cleanup
        registryService.init();

    }

    @Test
    public void testInitWithProvider() throws IOException {
        // Create provider structure under terraform/providers/
        Path registryPath = REGISTRY_STORAGE_PATH.resolve("terraform");
        Path providersPath = registryPath.resolve("providers");
        Path namespacePath = providersPath.resolve("hashicorp");
        Path providerPath = namespacePath.resolve("aws");
        Path versionPath = providerPath.resolve("1.0.0");
        Path platformPath = versionPath.resolve("linux").resolve("amd64");

        Files.createDirectories(platformPath);
        //Files.createFile(providersPath.resolve("namespaces.json"));

        // Re-init service after test data creation
        registryService.init();

        assertThat(registryService.getProviders()).hasSize(1);
        assertThat(registryService.getProvider("terraform", "hashicorp", "aws")).isPresent();
    }


    // Add more tests as needed
}

File: dotnet-impl/test/net/terria/registry/terraform/RegistryServiceTest.cs
using System;
using System.IO;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using NUnit.Framework;
using net.terria.registry.terraform;
using Microsoft.VisualBasic;

namespace net.terria.registry.terraform
{
    [TestFixture]
    public class RegistryServiceTest
    {

        private static string REGISTRY_STORAGE_PATH = "../../../../registry/";
        

        private String _registry_storage = REGISTRY_STORAGE_PATH;

        [SetUp]
        public void Setup()
        {
            if (!Directory.Exists(_registry_storage))
            {
                Directory.CreateDirectory(_registry_storage);
            }

             Console.Out.WriteLine("Registry Storage Root: " + Path.GetFullPath(_registry_storage));
        }

        [TearDown]
        public void TearDown()
        {
            // do nothing for now
            Console.Out.WriteLine("Registry Storage Root: " + Path.GetFullPath(_registry_storage));
        }

        [Test]
        public void TestInitWithModule()
        {
            // Create provider structure under terraform/providers
            var modulesPath = Path.Combine(_registry_storage, "terraform", "modules");
            var namespacePath = Path.Combine(modulesPath, "azure");
            var modulePath = Path.Combine(namespacePath, "terraform-azurerm-avm-utl-regions");
            var versionPath = Path.Combine(modulePath, "0.1.0");
            var platformPath = Path.Combine(versionPath, "azurerm");

            Directory.CreateDirectory(platformPath);

            var config = new ConfigurationBuilder()
                .AddInMemoryCollection(new[] { new KeyValuePair<string, string>("registry.storage.path", _registry_storage) })
                .Build();

            var logger = LoggerFactory.Create(builder => builder.AddConsole()).CreateLogger<RegistryService>();

            var service = new RegistryService(logger, config);

            Assert.That(service.GetModules().Count, Is.GreaterThan(0));
            Assert.That(service.GetModule("terraform", "azure", "terraform-azurerm-avm-utl-regions"), Is.Not.Null);
        }


        [Test]
        public void TestInitWithProvider()
        {
            // Create provider structure under terraform/providers
            var providersPath = Path.Combine(_registry_storage, "terraform", "providers");
            var namespacePath = Path.Combine(providersPath, "hashicorp");
            var providerPath = Path.Combine(namespacePath, "aws");
            var versionPath = Path.Combine(providerPath, "1.0.0");
            var platformPath = Path.Combine(versionPath, "linux", "amd64");

            Directory.CreateDirectory(platformPath);

            var config = new ConfigurationBuilder()
                .AddInMemoryCollection(new[] { new KeyValuePair<string, string>("registry.storage.path", _registry_storage) })
                .Build();

            var logger = LoggerFactory.Create(builder => builder.AddConsole()).CreateLogger<RegistryService>();

            var service = new RegistryService(logger, config);

            Assert.That(service.GetProviders().Count, Is.GreaterThan(0));
            Assert.That(service.GetProvider("terraform", "hashicorp", "aws"), Is.Not.Null);
        }

        // Add more tests as needed
    }
}